### DISCUSSION TITLE
PyVista vs vedo

### DISCUSSION THREAD

**Discussion by Sbozzolo** [pyvista/pyvista#3470] [Q&A]
I develop and maintain [kuibit](https://github.com/Sbozzolo/kuibit), a post-processing library for the [Einstein Toolkit](https://einsteintoolkit.org), a popular code for general relativistic astrophysics (think: neutron stars and black holes). `Einstein Toolkit` runs 3D simulations with a very simple mesh refinement and I would like to provide a module for some simple 3D visualization in `kuibit`.

Historically, the community used `VisIt` for 3D visualization, but I want something that can be easily installed and scripted (even if not as powerful). There's 3 types of renderings that we commonly make (for simplicity, we can assume that everything is defined on a regular uniform Cartesian grid):
* Volume rendering
* Field lines
* Simple meshes (that roughly will always look like spheres approximately, this is the black hole horizon)

This is an example of image that I would like to make:
![fig1](https://user-images.githubusercontent.com/9167485/195898155-aca295cc-80fb-4607-ab25-98806242d5e9.png)

So far, I've use `mayavi` as 3D plotting library for `kuibit`, but I increasingly got more and more frustrated with it. I liked that `mayavi` is simple and I could make the aforementioned plots easily, but it became a massive burden to support. Therefore, I am trying to explore alternatives. At this point, I think `PyVista` and `vedo` are the two main candidates. There is an [old discussion](https://github.com/pyvista/pyvista/issues/146) about comparing `PyVista` with other tools, but I am sure lots have changed, and I could not find updated information.

What are the most important differences between the two packages? Is `PyVista` a good tool for my use case?

**Comment by banesullivan** [comment]
Hi @Sbozzolo! I'm thrilled to hear your about your effort to improve the 3D plotting capabilities of kuibit! PyVista was built for exactly this - being a supporting module for analysis and 3D visualization. I'd like to take some time to thoughtfully craft a response to your question and sell you on why I think PyVista is a great choice 😉 but I'd also like to ping @marcomusy for a chance to chime in on this discussion.

> There is an old discussion about comparing PyVista with other tools, but I am sure lots have changed, and I could not find updated information.

Old indeed and definitely no longer captures the state of the technologies. PyVista and Vedo are both based on VTK but take different approaches to how they wrap and streamline usage of VTK -- I'll further explain in a more detailed post (hopefully in the next few days), to not only outline the differences for you, but re-capture the state of VTK-based 3D visualization in Python at this time for other's looking for insight.

**Reply by akaszynski** [reply]
Great question.

As the co-creator of PyVista, I'm going to be a bit biased, so I'll try to base this on facts instead of opinions or anything subjective. I'm only going to speak about the pros here, and any points that I missed are either due to my inexperience or inherent bias. Also, let it be clear: open source is awesome and we are all indebted to the work these individuals (really, the entire community) has put in to these projects. We all benefit from their hard work and I think any choice you make will be a good one because it will be the one that works for you.

### Pros for PyVista and Vedo (relative to mayavi)
- Popularity: Despite mayavi having a nearly 6 year head start (at least from the time it was on GitHub), both PyVista and vedo have gained significant traction. See [Star History](https://star-history.com/#pyvista/pyvista&marcomusy/vedo&enthought/mayavi&Date)
- Easy to install. Because both PyVista and Vedo are pure python, it's easy to install, and now that Kitware publishes VTK wheels to PyPI, it's really straightforward on any platform. It's a bit less so because of mayavi's additional build process and dependencies.

### Pros for Vedo:
- Looks gorgeous out of the box. I think @marcomusy has done a lot of tuning to configure sensible defaults that we're probably going to add to a mode to PyVista as `pyvista.theme.beautiful()`.
- Great examples that can be run from `vedo -r earthquake_browser`. Really, all the [examples](https://vedo.embl.es/) look fantastic. It's a great feature that lets you play around with vedo by writing a single line of code.
- Awesome 2d charts in a 3D plotter. The example I'm thinking of is `vedo -r extra_yaxis`.
- Excellent choice of fonts. PyVista is currently limited to the build-in VTK fonts and vedo's are much better, or at least friendlier looking. 
- [Docs](vedo -r extra_yaxis) are great, examples are awesome. Caveat: PyVista uses the PyData theme for consistency, but I personally think the docs are better. Guess that's a subjective opinion. I think our examples are good, but many of vedo's are just stunning.

### Pros for PyVista:
- Wider adoption in the Python community. [PyVista - PyPI download stats](https://pypistats.org/packages/pyvista) vs [vedo - PyPI download stats](https://pypistats.org/packages/vedo) indicates PyVista is ~ 8x more popular (58k vs 7.2k downloads in the past month as of 14 October 2022).
- [Used by more libraries](https://github.com/pyvista/pyvista/network/dependents). This is different than downloads. These are the repositories that depend on PyVista as reported by GitHub, and as of today, it's about 7x more for PyVista. Given that both libraries started at about the same time, it means that PyVista has had better adoption by other libraries given a relatively level playing field.
- [Data APIs](https://docs.pyvista.org/api/core/_autosummary/pyvista.DataSetAttributes.html#pyvista.DataSetAttributes). The way we access data has arisen from the need to have clean and efficient APIs to access the point, cell, and field data while making it convenient and easy. This is really important when building libraries off of PyVista + VTK, you need to be able to access and manipulate the data.
- [Plugins with Qt](https://github.com/pyvista/pyvistaqt). It's actually more important that you think, and it's used by a variety of closed source projects (including my own) and it makes it possible to ship fully featured standalone applications with `pyinstaller` that are quite awesome. In the future we plan on having better plugins with other frameworks like `trame` and `dash`, and there are a variety of examples out there on the web that already do this.
- Numpy/Scipy/Pandas-like docs with fully [fleshed examples](https://docs.pyvista.org/examples/index.html) as well as [micro examples](https://docs.pyvista.org/api/plotting/_autosummary/pyvista.Plotter.add_points.html#add-points). Again, this was a design choice and doesn't look as good as vedo's, but it's organized like people would expect it to be for a data scientist/engineer.

---

Again, I'd like to emphasize that all three libraries are solid choices, and mostly it's because VTK is so great as a base library. I'd urge you to play around with both PyVista's and vedo's APIs and get a feel for which one works for you, both for small simple scenes as well as complex ones. If you need any help implementing specific examples, [post an issue ](https://github.com/pyvista/pyvista/issues) and we'll try to help.

**Reply by Sbozzolo** [reply]
Thank you very much.

In the spirit of surveying the current state-of-the-art: what is drawing me to `PyVista` and `vedo` is the active and vibrant communities, and the level of maintainer's engagement. 

While I am indebted to `mayavi`, as a downstream developer, it is extremely important for me that my users can easily install my library. In multiple instances, `mayavi` was the only problematic component among my dependencies (+ we had to wait for one year for VTK and Python 3.10...), and basic issues with being able to install the library would not get any acknowledgement on the tracker (e.g., https://github.com/enthought/mayavi/issues/1063). This is the opposite of what I am seeing here.

(Thanks for producing and supporting amazing code!)

**Reply by marcomusy** [reply]
Thanks Alex and Bane for the nice review of `vedo`, and Gabriele for your interest!
One other thing in favour of `pyvista` is a better support for jupyter notebooks thanks to pythree (I had it in my own todo list for a long time but never found the right moment to implement it in `vedo`). About `pyinstaller` you can also use it in `vedo` to produce standalone executables.
As Alex said you can play a bit with the two and choose the flavour you like most :) There is also another one which is [fury](https://github.com/fury-gl/fury) which offers great support for custom shaders.

About the kind of renderings that you wish to achieve, the most look-alike example we have in `vedo` is this (computing the B-field from an electric wire interactively):

<img width="528" alt="Screenshot 2022-10-15 at 12 17 05" src="https://user-images.githubusercontent.com/32848391/195981896-95f94355-8c22-45a2-8f66-c993cf1e66fd.png">

this is what the code looks like ([click](https://github.com/marcomusy/vedo/blob/master/examples/simulations/mag_field1.py))
